src/Microsoft.OpenApi/Services/OpenApiWalker.cs

```diff
@@ -33,132 +35,186 @@ public void Walk(OpenApiDocument doc)
             _visitor.Visit(components);
-            _visitor.Exit();
+
+            if (components == null)
+            {
+                return;
+            }
+
+            if (components.Schemas != null)
+            {
+                foreach (var item in components.Schemas)
+                {
+                    InContext(item.Key, () => Walk(item.Value));
+                }
+            }
+
+            if (components.Callbacks != null)
+            {
+                foreach (var item in components.Callbacks)
+                {
+                    InContext(item.Key, () => Walk(item.Value));
+                }
+            }
+
+            if (components.Parameters != null)
+            {
+                foreach (var item in components.Parameters)
+                {
+                    InContext(item.Key, () => Walk(item.Value));
+                }
+            }
+
+            if (components.Examples != null)
+            {
+                foreach (var item in components.Examples)
+                {
+                    InContext(item.Key, () => Walk(item.Value));
+                }
+            }
+
+            if (components.Headers != null)
+            {
+                foreach (var item in components.Headers)
+                {
+                    InContext(item.Key, () => Walk(item.Value));
+                }
+            }
+
+            if (components.Links != null)
+            {
+                foreach (var item in components.Links)
+                {
+                    InContext(item.Key, () => Walk(item.Value));
+                }
+            }
+
+            if (components.RequestBodies != null)
+            {
+                foreach (var item in components.RequestBodies)
+                {
+                    InContext(item.Key, () => Walk(item.Value));
+                }
+            }
+
+            if (components.Responses != null)
+            {
+                foreach (var item in components.Responses)
+                {
+                    InContext(item.Key, () => Walk(item.Value));
+                }
+            }
+
+            Walk(components as IOpenApiExtensible);
         }
 
         /// <summary>
         /// Visits <see cref="OpenApiPaths"/> and child objects
         /// </summary>
-        /// <param name="paths"></param>
         internal void Walk(OpenApiPaths paths)
         {
-            _visitor.Enter(OpenApiConstants.Paths);
             _visitor.Visit(paths);
-            foreach (var pathItem in paths)
+
+            // Visit Paths
+            if (paths != null)
             {
-                _visitor.Enter(pathItem.Key.Replace("/","~1"));  // JSON Pointer uses ~1 as an escape character for /
-                Walk(pathItem.Value);
-                _visitor.Exit();
+                foreach (var pathItem in paths)
+                {
+                    InContext(pathItem.Key.Replace("/", "~1"), () => Walk(pathItem.Value));// JSON Pointer uses ~1 as an escape character for /
```

Comment:
This "Replace" should technically apply to anything in the JSON pointer (not just pathItem names). Can we move it to the InContext method?